cmd/gitcloner/gogit/cloner.go

```diff
@@ -0,0 +1,174 @@
+package gogit
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/transport"
+	httpgit "github.com/go-git/go-git/v5/plumbing/transport/http"
+	gossh "github.com/go-git/go-git/v5/plumbing/transport/ssh"
+	"github.com/rancher/gitjob/cmd/gitcloner/cmd"
+	giturls "github.com/whilp/git-urls"
+	"golang.org/x/crypto/ssh"
+)
+
+const defaultBranch = "master"
+
+var (
+	plainClone = git.PlainClone
+	readFile   = os.ReadFile
+)
+
+type Cloner struct{}
+
+type Options struct {
+	Repo            string
+	Branch          string
+	Auth            transport.AuthMethod
+	InsecureSkipTLS bool
+	CABundle        []byte
+}
+
+func NewCloner() *Cloner {
+	return &Cloner{}
+}
+
+func (c *Cloner) CloneRepo(opts *cmd.Options) error {
+	auth, err := createAuthFromOpts(opts)
+	if err != nil {
+		return err
+	}
+	caBundle, err := getCABundleFromFile(opts.CABundleFile)
+	if err != nil {
+		return err
+	}
+
+	if opts.Branch == "" && opts.Revison == "" {
+		opts.Branch = defaultBranch
+		return cloneBranch(opts, auth, caBundle)
+	}
+
+	if opts.Branch != "" {
+		return cloneBranch(opts, auth, caBundle)
```

Comment:
I don't know if it's intended but this implementation ignores the revision if a branch is specified, but they are not mutually exclusive. What about extracting the "checkout" mechanism into its own function and apply if if revision is not empty on either case?

Reply:
In Rancher UI you can just choose a branch or a revision (dropdown), so at least in the Rancher UI they are exclusive. I think it would be confusing to checkout the revision if a branch is provided, and also watch for changes in the branch? That's why I think revision should be ignored if branch is provided

Reply:
But note this is not Rancher nor Fleet. Also, `gitcloner` doesn't need to know about tracking, just cloning, right?
By doing this, we would be tying the design of gitjob/gitcloner to that specific use case.
In my opinion, if using branch and revision at the same time does not make sense for Fleet/Rancher, it should be the caller of `gitcloner` who needs to omit the revision from the GitJob CRD spec.

Reply:
To be honest nobody will use `gitcloner` apart from Fleet/Rancher, so I would avoid implementing something that is never going to be used. What if we just return an error in `gitcloner` if both `revision` an `branch` are provided? 
GitJob will never pass both parameters to `gitcloner` see [here](https://github.com/rancher/gitjob/blob/18d1936849a81d6c5a7b4bf17a331bb5eda08033/pkg/controller/gitjob/gitjobs.go#L251-L255)

Reply:
OK! Taking that into account, maybe printing a warning is enough, and reduces the chances of unintentionally breaking something.

Reply:
warning added https://github.com/rancher/gitjob/pull/302/commits/d4469f08ba84981ad82da69a7dcd9ca5cb8513eb